src/basic/cgroup-util.c

```diff
@@ -177,6 +177,24 @@ bool cg_kill_supported(void) {
         return supported;
 }
 
+bool cg_zswap_supported(void) {
+        static thread_local int supported = -1;
+
+        if (supported >= 0)
+                return supported;
+
+        if (cg_all_unified() <= 0)
+                supported = false;
+        else if (access("/sys/fs/cgroup/init.scope/memory.zswap.max", F_OK) < 0) {
```

Comment:
One thing I would change is using `cg_pid_get_path()` here and then using that to get the complete path to check the existence of `memory.zswap.max`.

Reply:
This check relies on memory controller being enabled for `init.scope` (which is technically correct, but it's a dynamic property and cache can have a stale incorrect value). (Also it relies on swap accounting enabled (not disabled) on kernel boot.)

I'd suggest simply ignoring ENOENT upon writes in a particular cgroup as with `memory.swap.max`.

Reply:
hmm, why isn't zswap availability reported in /sys/kernel/cgroup/features?

Reply:
(and could it be added there?)

Reply:
(The features file lists features that change semantics of the exposed files and match mount options. It could be changed but it'd be redundant to check of the attribute file existence.)

This is IMO simpler to approach the same way like `memory.swap.max` is already handled. (And devise more sophisticated approach if it proves slow.)